geospaas/base_viewer/views.py

```diff
@@ -56,55 +43,42 @@ def post(self, request, *args, **kwargs):
-        return self.sorting_func(request)
+
+        # modify attributes based on the forms cleaned data
+        self.filtering_func(request)
+        return self.final_rendering(request)
 
     def validation_and_error_check(self):
         for element_forms in self.forms:  # for loop for making clean data by is_valid() method
             element_forms.is_valid()
-            for errorField in element_forms.errors: # temporary error message
+            for errorField in element_forms.errors:  # temporary error message
                 print(f"errorField")
 
-    def sorting_func(self, request):
-        self.ds = CatalogDataset.objects.all()
-        self.filtering_func()
-        return self.final_rendering(request)
-
+    def filtering_func(self, request):
+        ds = CatalogDataset.objects.all()
+        if (request.method == 'POST'):
+            for counter in range(len(self.form_class)):
+                # using the filter function of each form sequentially
+                ds = self.forms[counter].filter(ds)
+            self.ds = ds
+        elif (request.method == 'GET'):
+            self.ds = ds
 
     def final_rendering(self, request):
         ''' Render page based on several forms of data as well as some other context '''
+        self.context = {}
         self.set_context()
         return render(request, self.main_template, self.context)
 
-
-# functions to be overloaded in child classes
-
-    def filtering_func(self):
-        # filter by source
-        src=[]
-        for element_forms in self.forms:
-            if 'source' in element_forms.cleaned_data:
-                if element_forms.cleaned_data['source'] is not None:
-                    src.append(element_forms.cleaned_data['source'])
-
-        self.ds = self.ds.filter(source__in=src)
-        # filter by time
-        t0 = self.forms[1].cleaned_data['time_coverage_start']
-        t1 = self.forms[1].cleaned_data['time_coverage_end'] + timezone.timedelta(hours=24)
-        self.ds = self.ds.filter(Q(time_coverage_end__lt=t1) & Q(time_coverage_start__gt=t0))
-        # filter by geometry
-        if self.forms[0].cleaned_data['polygon'] is not None:
-            self.ds = self.ds.filter(geographic_location__geometry__intersects= self.forms[0].cleaned_data['polygon'])
-        # ds = blabla(ds)
-        #return self.ds
-
     def set_context(self):
-        self.context = {}  # initializing the contect for templates
         # initializing the list of the forms that passed into context
         form_list = [None] * len(self.forms)
         for counter, value in enumerate(self.forms):
             # make the form_list elements one by one (form by form)
             form_list[counter] = value
         # passing the form_list into context
         self.context['form_list'] = form_list
+        self.set_dataset_context()
+
+    def set_dataset_context(self): # excluded from other context for pagination development of django
```

Comment:
Why do you need a separate function for just one line? 
```suggestion
```

Reply:
Please have a look at my comments as well!! I wrote that in the comment:

# excluded from other context for pagination development of django


In the possible child class we need to override this function if we want to take advantage of django pagination in the future.